add docs for command dedup (#11928)
digital-asset/daml: docs/source/app-dev/command-deduplication.rst

#. How command deduplication works.

#. How applications can effectively use the command deduplication.

.. _command-dedup-workings:

How command deduplication works
*******************************

The following fields in a command submissions are relevant for command deduplication.
The first three form the :ref:`change ID <change-id>` that identifies the intended ledger change.

#. The union of :ref:`party <com.daml.ledger.api.v1.Commands.party>` and :ref:`act_as <com.daml.ledger.api.v1.Commands.act_as>` define the submitting parties.
  
#. The :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` identifies the application that submits the command.

#. The :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>` is chosen by the application to identify the intended ledger change.

#. The :ref:`deduplication period <com.daml.ledger.api.v1.Commands.deduplication_period>` specifies the period for which no earlier submissions with the same change ID should have been accepted, as witnessed by a completion event on the :ref:`command completion service <command-completion-service>`.
   Otherwise, the current submission shall be rejected.

>    The period is specified either as a :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` or as a :ref:`deduplication offset <com.daml.ledger.api.v1.Commands.deduplication_offset>`.

Author: Good idea to make this explicit. Why did you pick `exclusive` though? The original design says `inclusive` (https://docs.google.com/document/d/1R0yAD2buzkHjPiWAHm-1FtX8acQofZLRGA4KDZ4NI2w/) and so do the docs for the ReadService: https://github.com/digital-asset/daml/blob/main/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/ReadService.scala#L83

>    The period is specified either as a :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` or as a :ref:`deduplication offset <com.daml.ledger.api.v1.Commands.deduplication_offset>`.

Reviewer: My misremembering, resolving.

> #. The :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` is returned to the submitting application and allows it to correlate specific submissions to specific completions.

Reviewer 2: ```suggestion
#. The :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` is chosen by the application to identify a specific submission. It is included in the corresponding completion event and allows the submitting application to correlate specific submissions to specific completions. An application should never reuse a submission ID.
```
if you want to make it more explicit that it's used in the completion stream

> #. The :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` is returned to the submitting application and allows it to correlate specific submissions to specific completions.

Author: Good point. I'll rephrase.

   An application should never reuse a submission ID.

> The ledger may arbitrarily extend the deduplication period specified in the submission, even beyond the maximum deduplication time specified in the :ref:`ledger configuration <ledger-configuration-service>`.

Author: As explained on in the protobuf, this is the maximum value that the participant/ledger promises to always handle. Since longer deduplication periods are strictly better than shorter ones, the application should be happy if it gets more guarantees.

> The ledger may arbitrarily extend the deduplication period specified in the submission, even beyond the maximum deduplication time specified in the :ref:`ledger configuration <ledger-configuration-service>`.

Reviewer: > As explained on in the protobuf, this is the maximum value that the participant/ledger promises to always handle.

I'd also state it here as it's an important guarantee and it would be nice to avoid having to jump to the protobuf, especially considering that this guide is already (and luckily so) very comprehensive.

> longer deduplication periods are strictly better than shorter ones

I wonder if there are use cases where the application would rely on deduplication not to exceed a certain time frame.

The deduplication period chosen by the ledger is the *effective deduplication period*.
The ledger may also convert a requested deduplication duration into an effective deduplication offset or vice versa.
The effective deduplication period is reported in the the command completion event in the :ref:`deduplication duration <com.daml.ledger.api.v1.Completion.deduplication_duration>` or :ref:`deduplication offset <com.daml.ledger.api.v1.Completion.deduplication_offset>` fields.

A command submission is considered a **duplicate submission** if at least one of the following holds:

- The submitting participant's completion service contains a successful completion event for the same :ref:`change ID <change-id>` within the *effective* deduplication period.

- The participant or Daml ledger are aware of another command submission in-flight with the same :ref:`change ID <change-id>` when they performs command deduplication.

> Command submissions via the :ref:`command service <command-service>` indicate the command deduplication outcome as a synchronous gRPC response (except when the gRPC deadline was exceeded before).

Author: Good idea. I've added a link to https://grpc.io/blog/deadlines/. Let me know if you have a better reference.

Submissions via the :ref:`command submission service <command-submission-service>` can indicate the outcome synchronously or asynchronously in the event on the :ref:`command completion service <command-completion-service>`.
In particular, the submission may be a duplicate even if the command submission service acknowledges the submission with the gRPC status code ``OK``.

> Command deduplication generates the following outcomes of a command submission:

Reviewer 2: It's unclear to me whether the following list only applies to completion events, or if the same errors can also be returned synchronously by the submission service.

Author: The sentences above this one say that outcomes are reported either synchronously or asynchronously. I've moved the sentence to the beginning of the paragraph. Does that help with the unclarity?

Author: Thanks for the suggested structure. I've added it, with a note saying that the synchronous completion of the command service may also appear on the completion stream (that's apparently something that some readers of our docs were surprised to find out).

- If no conflicting submission with the same :ref:`change ID <change-id>` was found, the completion event and the response convey the result of the submission (success or a gRPC error).

> - The gRPC status code ``ALREADY_EXISTS`` with error code id :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>` if there is an earlier command completion for the same :ref:`change ID <change-id>` within the effective deduplication period.

Author: At least we capitalize ID in many other places too.

- The gRPC status code ``ABORTED`` with error code id :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>` if another submission for the same :ref:`change ID <change-id>` was in flight when this submission was processed.

- The gRPC status code ``FAILED_PRECONDITION`` with error code id :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>` if the specified deduplication period is not supported.

>   The fields ``longest_duration`` or ``earliest_offset`` in the metadata specify the longest duration or earliest offset that is currently supported on the Ledger API endpoint.

Author: > For the metadata, we use max_deduplication_duration to represent the maximum allowed duration.

Why did you pick `max_deduplication_duration` as the field name? I thought that we had settled on `longest_duration` in the UX design doc and that's what I have implemented in Canton. BTW Canton may return durations in `longest_duration` that are different from the `max_deduplication_time` in the ledger config.

  At least one of the two fields is present.

>   Deduplication periods that span at most the :ref:`maximum deduplication time <com.daml.ledger.api.v1.LedgerConfiguration.max_deduplication_time>` SHOULD not result in this error.

Author: Added.

>   Participants may accept longer periods at their discretion.

Reviewer 2: I agree. The current documentation makes it sound like the maximum is useless:

1. Users can supply periods longer than the "maximum"
2. The participant can extend periods beyond the "maximum"

I assume (2) is because of technical reasons the participant/ledger may have to extend the period by a small duration.
What is the reason for (1)? Validating the duration is a bit annoying because the participant doesn't know the latest ledger configuration and there could be race conditions. However, the ledger should be able to validate it the same way it validates the ledger time model (which also depends both on the current configuration and the submitted transaction)

>   Participants may accept longer periods at their discretion.

Author: I agree that `maximum` is a misleading name here. I'm happy to go with a different name. The reason why it's still called `maximum deduplication time` is because that's the name of the field in the ledger configuration protobuf message and we didn't dare to rename it (yet).

>   Participants may accept longer periods at their discretion.

Reviewer: IIUC it's really the "maximum deduplication time guaranteed to be supported by the ledger / operator" but that's very verbose and I am not sure how it could be shortened. Also, a naming misalignment with the proto would indeed be a problem.

>   Participants may accept longer periods at their discretion.

Author: > I assume (2) is because of technical reasons the participant/ledger may have to extend the period by a small duration.

Not only for technical reasons. If the ledger reports a longer dedup period, then the application gains more knowledge. So longer dedup periods are in a sense better.

> What is the reason for (1)?

The reasoning is as follows: If it is no effort for the participant / ledger to support longer dedup period, then this is something we should make available to applications on a best-effort bases. They shouldn't rely on it in general. But suppose we are in the unforeseen eventuality that an outage has lasted for long enough that the max deduplication time is too short (especially when a deduplication offset is used). Then it just feels wrong to say "sorry, I can't help you with this because your deduplication offset is too early" even though it can be supported without problems because the participant happens to still have the data available.

>   Participants may accept longer periods at their discretion.

Reviewer 2: > maximum deduplication time guaranteed to be supported by the ledger / operator

I like this explanation (and also the longer explanation by @andreaslochbihler-da helped clearing it up for me).

> For deduplication to work as intended, all submissions for the same ledger change must be submitted via the same participants.

Reviewer 2: ```suggestion
For deduplication to work as intended, all submissions for the same ledger change must be submitted via the same participant.
```

Author: Changed.

> This is because a participant outputs by default only the completion events for submissions that were requested via this participant;

Author: Correct, the participant doesn't know this in general and (at least for Canton) this is fundamentally so. I don't fully understand in what direction you want to improve this? The wording or the fact that we don't provide cross-participant deduplication?

As a typical example, the application wants to create exactly one contract instance of template ``T`` via the corresponding :ref:`Create command <com.daml.ledger.api.v1.CreateCommand>`.
We assume that the ledger is correctly set up so that the Create command should be accepted, i.e., Daml packages are uploaded and vetted, the parties are allocated, and the application has a valid access token for the submission.

Known processing time bounds
============================

For now, we assume that the application knows a bound ``B`` on the processing time and forward clock drifts (against the application’s clock) in the Daml ledger.
If processing measured across all retries takes longer than the bound ``B``, the ledger change may take effect several times.
Under this assumption, the following strategy works for applications that use the :ref:`Command Service <command-service>` or the :ref:`Command Submission <command-submission-service>` and :ref:`Command Completion Service <command-completion-service>`.

.. _dedup-bounded-step-command-id:

> #. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
>    Remember the command ID across application crashes.

Reviewer 2: Isn't using deterministic command IDs an alternative to remembering the command ID across application restarts? The text makes it sound like you always have to remember command IDs.

```suggestion
#. Choose a command ID for the ledger change, in a way that makes sure the same ledger change is always assigned the same command ID. Either determine the command ID deterministically (e.g., if your contract payload contains a globally unique identifier, you can use that as your command ID), or choose command IDs randomly and remember them across application crashes.
```
```

Author: That's a much better phrasing. If you use a deterministic scheme, then you can remember the command ID by remembering the scheme. But that fine point is probably not obvious in my version..

>    Remember the command ID across application crashes.

Author: If the application were to use a random UUID for the command ID to identify the ledger change, then the application must persist that it has picked a given UUID for some fixed intended ledger change so that if the application crashes and restarts, it will use the same UUID again for resubmissions and not pick a fresh UUID. That's why we recommend to derive the command ID deterministicly.

>    Remember the command ID across application crashes.

Reviewer: Oh I totaly misunderstood (my brain picked `crashes` as a verb for some reason). Perhaps expanding a bit can help, e.g. `Persist the command ID so that, if the application crashes and restarts, it will use the same command ID for resubmissions`.

>    Remember the command ID across application crashes.

Author: Expanded.

   .. note::
      Make sure that you set command ID deterministically, that is to say: the same ledger change must use the same command ID for all command submissions.
      This is useful for the recovery procedure after an application crash/restart, in which the application inspects the state of the ledger (e.g., via the :ref:`Active contracts service <active-contract-service>`) and sends commands to the ledger.

>       When using deterministic command IDs, any commands that had been sent before the application restart will be discarded by the ledger to avoid duplicate submissions.

Reviewer 2: The ledger doesn't know about application restarts. By using deterministic command IDs, your application will automatically work correctly even after crash recovery. You just resubmit all commands after recovery without having to robustly track in-flight commands, and the ledger will take care of deduplicating them.

>       When using deterministic command IDs, any commands that had been sent before the application restart will be discarded by the ledger to avoid duplicate submissions.

Author: The ledger doesn't have to know this. It just deduplicates things. The problem is that the application cannot know after a crash whether it has submitted a command or not. I've rephrased this whole paragraph. Please have a look whether it's clearer now.

>       When using deterministic command IDs, any commands that had been sent before the application restart will be discarded by the ledger to avoid duplicate submissions.

Reviewer: I think it is much clearer, thanks!

   .. _dedup-bounded-step-offset:

#. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.

   .. _dedup-bounded-step-submit:
   
#. Submit the command with the following parameters

   - Set the :ref:`command ID <<com.daml.ledger.api.v1.Commands.command_id>>` to the chosen command ID from :ref:`step 1 <dedup-bounded-step-command-id>`.

>    - Set the :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` to the bound ``B``.

Author: Not sure. This is meant as a recipe for application writers, so the fewer choices they have to make, the better. What's your motivation for changing it to >= B?

>    - Set the :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` to the bound ``B``.

Author: That's the same question as why would anyone specify a deduplication duration for command deduplication on the ledger API. I'll put a `note` in here that the default deduplication duration is the max dedup time and so they may just not set it at all.

   - Set the :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` to a fresh value, e.g., a random UUID.

   - Set the timeout (gRPC deadline) to the expected processing (Command Service) or submission (Command Submission Service) delay.

     The **processing delay** measures the time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives the acceptance or rejection.
     The **submission delay** measures the time between when the application sends off a submission to the :ref:`Command Submission Service <command-submission-service>` and when it obtains a synchronous response for this gRPC call.
     After the RPC timeout, the application considers the submission as lost and enters a retry loop.

>      This timeout is typically much shorter than the deduplication duration.

Author: Yes. You get some gRPC response back from the asynchronous command submission service as well.

>    - Set the timeout (gRPC deadline) to the expected processing (Command Service) or submission (Command Submission Service) delay.
> 
>      The **processing delay** measures the time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives the acceptance or rejection.
>      The **submission delay** measures the time between when the application sends off a submission to the :ref:`Command Submission Service <command-submission-service>` and when it obtains a synchronous response for this gRPC call.
>      After the RPC timeout, the application considers the submission as lost and enters a retry loop.
>      This timeout is typically much shorter than the deduplication duration.

Reviewer: I think this would clear my confusion:

```suggestion
   - Set the timeout (gRPC deadline) to the expected processing (Command Service) or submission (Command Submission Service) delay.

     The **processing delay** measures the submission processing time, i.e. the time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives (synchronously, unless it times out) the acceptance or rejection.
     The **submission delay** measures the submission hand-off time, i.e. the time between when the application sends off a submission to the :ref:`Command Submission Service <command-submission-service>` and when it obtains a synchronous response for this gRPC call.
     After the RPC timeout, the application considers the submission as lost and enters a retry loop.
     This timeout is typically much shorter than the deduplication duration.
```

   
#. Wait until the RPC call returns a response.
   
   - Status codes other than ``OK`` trigger :ref:`error handling <dedup-bounded-error-handling>`.

   - When you use the :ref:`Command Service <command-service>` and the response carries the status code ``OK``, the ledger change took place.
     You can report success.
     
   - When you use the :ref:`Command Completion Service <command-completion-service>`,
     subscribe with the :ref:`Command Completion Service <command-submission-service>` for completions for ``actAs`` from ``OFF1`` exclusive until you see a completion event for the change ID and the submission ID chosen in :ref:`step 3 <dedup-bounded-step-submit>`.
     If the completion’s status is ``OK``, the ledger change took place and you can report success.
     Other status codes trigger :ref:`error handling <dedup-bounded-error-handling>`.
   
     This step needs no timeout as the :ref:`Command Submission Service <command-submission-service>` acknowledges a submission only if there will eventually be a completion event, unless relevant parts of the system become permanently unavailable.

.. _dedup-bounded-error-handling:

Error handling
--------------

> Error handling is needed when the status code of the RPC call or in the :ref:`in the completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.

Reviewer 2: ```suggestion
Error handling is needed when the status code of the command submission gRPC call or in the :ref:`in the completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.
```

"the RPC call" is a bit ambiguous

The following table lists appropriate reactions by status code (written as ``STATUS_CODE``) and error code (written in capital letters with a link to the error code documentation).
Fields in the error metadata are written as ``field`` in lowercase letters.

.. list-table:: Command deduplication error handling with known processing time bound
   :widths: 10 50
   :header-rows: 1

   - * Error condition
     
     * Reaction

       
   - * ``DEADLINE_EXCEEDED``
     
     * Consider the submission as lost.
       

>        Retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`) and possibly increase the timeout.

Author: 1. If you use a duration, then the bound `B` includes all retries by assumption. So why should you?
2. If you use an offset, then you use the same deduplication offset again (`OFF0`). This automatically lengthens the deduplication period.

       
   - * Application crashed
     
     * Retry from obtaining a completion offset (:ref:`step 2 <dedup-bounded-step-offset>`).

   - * ``ALREADY_EXISTS`` / :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>`
     
     * The change ID has already been accepted on the ledger within the reported deduplication period.

>        The optional field ``completion_offset`` contains the precise offset.

Author: Added. :-)

       
   - * ``FAILED_PRECONDITION`` / :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>`
     
     * The specified deduplication duration is longer than what the Daml ledger supports.
       ``earliest_offset`` contains the earliest deduplication offset or ``longest_duration`` contains the longest deduplication duration that can be used (at least one of the two must be provided).

       Options:

       - Negotiate support for longer deduplication periods with the ledger operator.

       - Set the deduplication offset to ``earliest_offset`` or the deduplication duration to ``longest_duration`` and retry from obtaining a completion offset (:ref:`step 2 <dedup-bounded-step-offset>`).
	 This may lead to accepting the change twice.

	 
   - * ``ABORTED`` / :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>`
     
       This error occurs only as an RPC response, not inside a completion event.
       
     * There is already another submission in flight, with the submission ID in ``existing_submission_id``.

>        - When you use the :ref:`Command Service <command-service>`, wait a bit and retry from submitting the command (:ref:`step 3 <dedup-bounded-step-submit>`).

Author: It is not clear that the submission has already been accepted; it just says that there is an in-flight submission, but this submission could still be rejected. This can also happen with the command service. For example:
1. Submit via the command service with a very short gRPC deadline. The command service responds with `DEADLINE_EXCEEDED`.
2. You immediately resubmit with a higher deadline. This gives you `SUBMISSION_ALREADY_IN_FLIGHT`, as the submission from step 1 is still being processed. (This probably doesn't happen with kv based ledgers where the command deduplication check is atomic with the commit, but it can happen in Canton where the two are not atomic.)
3. Since the gRPC call from step 1 has exceeded the deadline, you don't know the outcome of the submission. So you wait a bit until you have some confidence that the earlier submission should now be processed. Alas, you don't want to use the completion service. So instead you submit the command again. If the submission from step 1 was accepted, then you'll get back ALREADY_EXISTS. If it failed, you now have a second chance to get it accepted by the ledger and learn the outcome.

       - When you use the :ref:`Command Completion Service <command-completion-service>`, look for a completion for ``existing_submission_id`` instead of the chosen submission ID in :ref:`step 4 <dedup-bounded-step-await>`.

   - * ``ABORTED`` / other error codes
     
     * Wait a bit and retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`).

>        Use the retryability information in the error details to decide how long to wait.

Author: Good point. I haven't found anything in the self-service error code design doc that `RetryInfo` is reliable. I'll remove the sentence.

       
   - * other error conditions

     * You should use background knowledge about the workflow to decide whether earlier submissions might still get accepted.
       If not, you may stop retrying and report that the ledger change failed.

>        If in doubt, retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`) or give up without knowing for sure that the ledger change will not happen.

Author: How is this different from the `background knowledge` I mentioned two sentences earlier?

>        If in doubt, retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`) or give up without knowing for sure that the ledger change will not happen.

Reviewer: Right, I am not sure about the helpfulness of these "If not" and "if in doubt" sentences though, i.e. I expect the best possible behavior (even if not completely resolutive) to be derived from the workflow analysis.

>        If in doubt, retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`) or give up without knowing for sure that the ledger change will not happen.

Author: I think I see where your confusion is coming from. The `If not` / `If in doubt` refer to the outcome of the workflow analysis. I'll make this clearer.

       In the running example of creating a contract instance of ``T``, you can never be sure, as any outstanding submission might still be accepted on the ledger.

>        In particular, you must not draw any conclusions from not having received a :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>` error, because the outstanding submission may be queued somewhere and will reach the relevant processing point only later.

Reviewer: True (but poor application).

>        In particular, you must not draw any conclusions from not having received a :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>` error, because the outstanding submission may be queued somewhere and will reach the relevant processing point only later.

Reviewer: I am even starting to think that we should publicly document a (logic/abstract) Daml processing pipeline (that ledgers could specialize / enrich) and talk about deduplication and deduplication errors (and errors in general) in that framework. I say that without taking away anything from this most excellent and very much needed command dedup compendium.

Failure scenarios
-----------------

The above strategy can fail in the following scenarios:

#. The bound ``B`` is too low: The command can be executed multiple times.
   
   Possible causes:

   - You have retried for longer than the deduplication duration, but never got a meaningful answer, e.g., because the timeout (gRPC deadline) is too short.
     For example, this can happen due to long-running Daml interpretation when using the :ref:`Command Service <command-service>`.

   - The application clock drifts significantly from the participant's or ledger's clock.

   - There are unexpected network delays.

>    - Submissions are retried internally in the participant or Daml ledger and those retries do not stop before ``B`` is over.

Author: Ideally, the ledger documentation should explain such things and a participant operator should inform its users about such parameters. For example, suppose that the kv write service internally retries sending a submission to the kv committer. Where are these retry parameters documented?

   Remember the command ID across application crashes.
   (Analogous to :ref:`step 1 above <dedup-bounded-step-command-id>`)

   .. _dedup-unbounded-step-dedup-offset:
   
#. Obtain a recent offset ``OFF0`` on the completion event stream and remember across crashes that you use ``OFF0`` with the chosen command ID.

   - When you use the :ref:`Command Service <command-service>`, to obtain a recent offset, repeatedly submit a dummy command, e.g., a :ref:`Create-And-Exercise command <com.daml.ledger.api.v1.CreateAndExerciseCommand>` of some single-signatory template with the :ref:`Archive <function-da-internal-template-functions-archive-52202>` choice, until you get a successful response.
     The response contains the :ref:`completion offset <com.daml.ledger.api.v1.SubmitAndWaitForTransactionIdResponse.completion_offset>`.

   - When you use the :ref:`Command Completion Service <command-completion-service>`, ask for the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.

   .. _dedup-unbounded-step-offset:

> #. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say its current end.

Author: No. This step can be skipped if you don't want to use the submission and completion services.

   (Analogous to :ref:`step 2 above <dedup-bounded-step-offset>`)

   .. note::

>       Unless error handling retries from this step, you can use ``OFF1 = OFF0``.

Author: Error handling instructs the application developer to `retry from step 3`. In that case, you want to obtain a fresh offset `OFF1`. If you execute this step the first time, you can use `OFF0`, because if you request another ledger end, you'll most likely get `OFF0` again. I've rephrased this now.

#. Submit the command with the following parameters (analogous to :ref:`step 3 above <dedup-bounded-step-submit>` except for the deduplication period):

   - Set the :ref:`command ID <<com.daml.ledger.api.v1.Commands.command_id>>` to the chosen command ID from :ref:`step 1 <dedup-bounded-step-command-id>`.

   - Set the :ref:`deduplication offset <com.daml.ledger.api.v1.Commands.deduplication_offset>` to ``OFF0``.

   - Set the :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` to a fresh value, e.g., a random UUID.

   - Set the timeout (gRPC deadline) to the expected processing (Command Service) or submission (Command Submission Service) delay.

#. Wait until the RPC call returns a response.
   

>    - Status codes other than ``OK`` trigger :ref:`error handling <dedup-bounded-error-handling>`.

Reviewer: ```suggestion
   - Status codes other than ``OK`` should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.
```

   - When you use the :ref:`Command Service <command-service>` and the response carries the status code ``OK``, the ledger change took place.
     You can report success.
     The response contains a :ref:`completion offset <com.daml.ledger.api.v1.SubmitAndWaitForTransactionIdResponse.completion_offset>` that you can use in :ref:`step 2 <dedup-unbounded-step-dedup-offset>` of later submissions.
     
   - When you use the :ref:`Command Completion Service <command-completion-service>`,
     subscribe with the :ref:`Command Completion Service <command-submission-service>` for completions for ``actAs`` from ``OFF1`` exclusive until you see a completion event for the change ID and the submission ID chosen in :ref:`step 3 <dedup-bounded-step-submit>`.
     If the completion’s status is ``OK``, the ledger change took place and you can report success.

>      Other status codes trigger :ref:`error handling <dedup-bounded-error-handling>`.

Reviewer: ```suggestion
     Other status codes should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.
```

   

Error handling
--------------

The same as :ref:`for known bounds <dedup-bounded-error-handling>`, except that the former retry from :ref:`step 2 <dedup-bounded-step-offset>` becomes retry from :ref:`step 3 <dedup-unbounded-step-offset>`.

Failure scenarios
-----------------

The above strategy can fail in the following scenarios:

#. No success within the supported deduplication period
   

>    When the application receives a :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>` error, it cannot achieve exactly once execution any more.

Reviewer: ```suggestion
   When the application receives a :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>` error, it cannot achieve exactly once execution any more within the originally intended deduplication period.
```

.. Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _command-deduplication:

Command deduplication
#####################

The interaction of a Daml application with the ledger is inherently asynchronous: applications send commands to the ledger, and some time later they see the effect of that command on the ledger.

> Many things can fail during this time window: the application can crash, the participant node can crash, messages can be lost on the network, or the ledger may be just slow to respond due to a high load.

Reviewer 3: "ledger may be just slow to load" - the word "just" is not needed

Author: Removed.

If you want to make sure that an intended ledger change is not executed twice, your application needs to robustly handle these failure scenarios.
This guide covers the following topics:

> #. How command deduplication works.

Reviewer 3: Should be an unordered list

Author: Changed.

#. How applications can effectively use the command deduplication.

.. _command-dedup-workings:

How command deduplication works
*******************************

The following fields in a command submissions are relevant for command deduplication.
The first three form the :ref:`change ID <change-id>` that identifies the intended ledger change.

#. The union of :ref:`party <com.daml.ledger.api.v1.Commands.party>` and :ref:`act_as <com.daml.ledger.api.v1.Commands.act_as>` define the submitting parties.
  

> #. The :ref:`application ID <com.daml.ledger.api.v1.Commands.application_id>` identifies the application that submits the command.

Reviewer 3: This should probably be an unordered list. Generally reserve ordered lists for action steps.

Author: I chose an ordered list because the paragraph before referred to the first three items. But you're right, numbering the items doesn't add any clarity. Changed to unordered.

#. The :ref:`command ID <com.daml.ledger.api.v1.Commands.command_id>` is chosen by the application to identify the intended ledger change.

> #. The :ref:`deduplication period <com.daml.ledger.api.v1.Commands.deduplication_period>` specifies the period for which, in order for the current submission to be accepted, no earlier submissions with the same change ID should have been accepted, as witnessed by a completion event on the :ref:`command completion service <command-completion-service>`.

Author: That's much simpler. Thanks.

   Otherwise, the current submission shall be rejected.
   The period is specified either as a :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` or as a :ref:`deduplication offset <com.daml.ledger.api.v1.Commands.deduplication_offset>` (inclusive).

#. The :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` progapates into the completion events so that the application can correlate specific submissions to specific completions.
   An application should never reuse a submission ID.

The ledger may arbitrarily extend the deduplication period specified in the submission, even beyond the maximum deduplication time specified in the :ref:`ledger configuration <ledger-configuration-service>`.

.. note::

>    The maximum deduplication time specifies a lower bound on the length of deduplication periods that the ledger accepts.

Reviewer 3: Can probably strike "that the ledger accepts"?

Author: Without the qualification, it could be misread as "shorter deduplication periods" are not allowed at all. I've phrased it differently now.

   
The deduplication period chosen by the ledger is the *effective deduplication period*.
The ledger may also convert a requested deduplication duration into an effective deduplication offset or vice versa.

> The effective deduplication period is reported in the the command completion event in the :ref:`deduplication duration <com.daml.ledger.api.v1.Completion.deduplication_duration>` or :ref:`deduplication offset <com.daml.ledger.api.v1.Completion.deduplication_offset>` fields.

Reviewer 3: duplicate "the" between "in" and "command"

A command submission is considered a **duplicate submission** if at least one of the following holds:

- The submitting participant's completion service contains a successful completion event for the same :ref:`change ID <change-id>` within the *effective* deduplication period.

- The participant or Daml ledger are aware of another command submission in-flight with the same :ref:`change ID <change-id>` when they perform command deduplication.

Command deduplication generates the following outcomes of a command submission.
Command submissions via the :ref:`command service <command-service>` indicate the command deduplication outcome as a synchronous gRPC response unless when the `gRPC deadline <https://grpc.io/blog/deadlines/>`_ was exceeded.
Submissions via the :ref:`command submission service <command-submission-service>` can indicate the outcome synchronously or asynchronously in the event on the :ref:`command completion service <command-completion-service>`.
In particular, the submission may be a duplicate even if the command submission service acknowledges the submission with the gRPC status code ``OK``.

> - If no conflicting submission with the same :ref:`change ID <change-id>` was found, the completion event and possibly the response convey the result of the submission (success or a gRPC error).

Reviewer 3: "was found" by who or what?

Author: Changed.

> - The gRPC status code ``ALREADY_EXISTS`` with error code ID :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>` if there is an earlier command completion for the same :ref:`change ID <change-id>` within the effective deduplication period.

Reviewer 3: No verb. Possibly "is displayed" or "is returned"? Parallel construction for all bullet points.

Author: Replaced `if` by `indicates that`, also for the other items in this list.

- The gRPC status code ``ABORTED`` with error code id :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>` if another submission for the same :ref:`change ID <change-id>` was in flight when this submission was processed.

- The gRPC status code ``FAILED_PRECONDITION`` with error code id :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>` if the specified deduplication period is not supported.
  The fields ``longest_duration`` or ``earliest_offset`` in the metadata specify the longest duration or earliest offset that is currently supported on the Ledger API endpoint.
  At least one of the two fields is present.

>   Neither deduplication durations up to the :ref:`maximum deduplication time <com.daml.ledger.api.v1.LedgerConfiguration.max_deduplication_time>` nor deduplication offsets published within that time SHOULD not result in this error.

Reviewer 3: This sentence appears to be a double negative as currently constructed. Clarify?

Author: You're right. Deleted "not". The negation is expressed already by "neither ... nor ...".

  Participants may accept longer periods at their discretion.

For deduplication to work as intended, all submissions for the same ledger change must be submitted via the same participant.

> This is because a participant outputs by default only the completion events for submissions that were requested via the very same participant,

Reviewer 3: "This is because a participant outputs by default only the completion events for submissions that were requested via the very same participant,
and whether a submission is considered a duplicate is determined by the completion events." change to "Whether a submission is considered a duplicate is determined by completion events, and by default a participant outputs only their own completion events."

Author: I took over the rephrasing with `whether`, but kept the ` the completion events for submissions that were requested via the very same participant`. That's because participants don't own completion events; at least I'm not aware that we use that terminology anywhere, just like participants do not own submissions. The submission merely passes through the participant, just like the completion.

and whether a submission is considered a duplicate is determined by the completion events.

> On some ledgers, every participant outputs the completion events for all the hosted parties, not just for the submissions that went through the participant.

Reviewer 3: Change "On some ledgers, every participant outputs the completion events for all the hosted parties, not just for the submissions that went through the participant. In this case, command deduplication works across participants. At this time, only Daml Driver for VMware Blockchain supports command deduplication across participants." to "At this time only Daml Driver for VMware Blockchain supports command deduplication across participants."

Author: Changed.

In this case, command deduplication works across participants.
At this time, only `Daml Driver for VMware Blockchain <https://www.digitalasset.com/daml-for-vmware-blockchain/>`__ supports command deduplication across participants.

    

How to use command deduplication
********************************

> When an application wants to effectuate a ledger change exactly once, the application must resubmit a command if an earlier submission was lost.

Reviewer 3: This sentence can be simplified to: "To make a one-time ledger change, the application must resubmit lost submissions."

Author: I'd like to keep `exactly once` because that's a technical term. But I've removed redundant the `When an application wants `.

> However, the application typically cannot distinguish a lost submission from a Daml ledger that is just slow in processing the submissions.

Reviewer 3: "However, the application typically cannot distinguish a lost submission from slow submission processing by the ledger."

Author: Changed.

> So the application should just resubmit the command until it is executed against the ledger, and rely on command deduplication to reject all duplicate submissions that do make it to the ledger.

Reviewer 3: "Command deduplication allows the application to resubmit the command until it is executed and reject all duplicate submissions thereafter."

Author: Changed. I feel that there's some subtlety lost, namely that an earlier (from the application perspective) submission may also be rejected by command deduplication, namely if it was overtaken by a later submission on its way to the participant or while the participant was processing it. But that's probably too much detail here.

> Some ledger changes can be executed at most once anyway, so no command deduplication is needed for them.

Reviewer 3: remove "anyway"

Author: Done.

All duplicate submissions of such a change will be rejected with :ref:`CONTRACT_NOT_ACTIVE <error_code_CONTRACT_NOT_ACTIVE>`.

In contrast, a :ref:`Create command <com.daml.ledger.api.v1.CreateCommand>` would create a fresh contract instance of the given :ref:`template <com.daml.ledger.api.v1.CreateCommand.template_id>` for each submission that reaches the ledger (unless other constraints such as the :ref:`template preconditions <daml-ref-preconditions>` or contract key uniqueness are violated).
Similarly, an :ref:`Exercise command <com.daml.ledger.api.v1.ExerciseCommand>` on a non-consuming choice or an :ref:`Exercise-By-Key command <com.daml.ledger.api.v1.ExercisebyKeyCommand>` may be executed multiple times if submitted multiple times.
With command deduplication, applications can ensure such intended ledger changes are executed only once within the deduplication period, even if the application resubmits, say because it considers the earlier submissions to be lost or forgot during a crash that it had already submitted the command.

> As a typical example, the application wants to create exactly one contract instance of template ``T`` via the corresponding :ref:`Create command <com.daml.ledger.api.v1.CreateCommand>`.

Reviewer 3: I think this is sufficiently clear without the example.

Author: Removed.

We assume that the ledger is correctly set up so that the Create command should be accepted, i.e., Daml packages are uploaded and vetted, the parties are allocated, and the application has a valid access token for the submission.

Known processing time bounds
============================

> For now, we assume that the application knows a bound ``B`` on the processing time and forward clock drifts in the Daml ledger w.r.t. the application’s clock.

Reviewer 3: spell out "with regard to"

Reviewer 3: Also, to the degree that this is user documentation it probably not wise to introduce thought experiments and assumptions that aren't universally valid.

Author: Spelled out.

I'm not sure why you call this a thought experiment. In a distributed system, you cannot get exactly once semantics unless you're willing to make some assumptions, e.g., how long it takes to process something. I've rephrased it to
```
For this strategy, you must estimate a bound ``B`` on the processing time and forward clock drifts in the Daml ledger with respect to the application’s clock.
```

   Remember the chosen command ID for the ledger change across application crashes.

   .. note::
      Make sure that you assign the same command ID to all command (re-)submissions of the same ledger change.
      If you derive the command ID deterministically from the ledger change, you must only remember the ledger change, but not the command ID because you can always derive the command ID.

      This is useful for the recovery procedure after an application crash/restart.
      After a crash, the application in general cannot know whether it has submitted a set of commands before the crash.
      If in doubt, resubmit the commands using the same command ID.
      If the commands had been submitted before the crash, command deduplication on the ledger will reject the resubmissions.

   .. _dedup-bounded-step-offset:

#. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.

   .. _dedup-bounded-step-submit:
   

> #. Submit the command with the following parameters

Reviewer 3: needs a colon at the end of the sentence to introduce the list of parametes

Author: Added.

     If the completion’s status is ``OK``, the ledger change took place and you can report success.
     Other status codes should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.
   
     This step needs no timeout as the :ref:`Command Submission Service <command-submission-service>` acknowledges a submission only if there will eventually be a completion event, unless relevant parts of the system become permanently unavailable.

.. _dedup-bounded-error-handling:

Error handling
--------------

Error handling is needed when the status code of the command submission RPC call or in the :ref:`in the completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.
The following table lists appropriate reactions by status code (written as ``STATUS_CODE``) and error code (written in capital letters with a link to the error code documentation).
Fields in the error metadata are written as ``field`` in lowercase letters.

.. list-table:: Command deduplication error handling with known processing time bound
   :widths: 10 50
   :header-rows: 1

   - * Error condition
     
     * Reaction

       
   - * ``DEADLINE_EXCEEDED``
     

>      * Consider the submission as lost.

Reviewer 3: Delete the word "as".

Author: Removed.

       
       Retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`) and possibly increase the timeout.

       
   - * Application crashed
     

>      * Retry by restarting the submission process from "obtaining a completion offset" (:ref:`step 2 <dedup-bounded-step-offset>`).

Author: Changed everywhere. I've also capitalized "Step" in all references.

       
   - * ``FAILED_PRECONDITION`` / :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>`
     
     * The specified deduplication period is longer than what the Daml ledger supports.
       ``earliest_offset`` contains the earliest deduplication offset or ``longest_duration`` contains the longest deduplication duration that can be used (at least one of the two must be provided).

       Options:

       - Negotiate support for longer deduplication periods with the ledger operator.

       - Set the deduplication offset to ``earliest_offset`` or the deduplication duration to ``longest_duration`` and retry from obtaining a completion offset (:ref:`step 2 <dedup-bounded-step-offset>`).
	 This may lead to accepting the change twice within the originally intended deduplication period.

	 
   - * ``ABORTED`` / :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>`
     
       This error occurs only as an RPC response, not inside a completion event.
       
     * There is already another submission in flight, with the submission ID in ``existing_submission_id``.

>        - When you use the :ref:`Command Service <command-service>`, wait a bit and retry from submitting the command (:ref:`step 3 <dedup-bounded-step-submit>`).

Reviewer 3: "retry from Step 2" as above.

Author: Done.

> 	 Since the in-flight submission might still be rejected, (repeated) resubmission ensures that you (eventually) learns the outcome:

Reviewer: ```suggestion
	 Since the in-flight submission might still be rejected, (repeated) resubmission ensures that you (eventually) learn the outcome:
```

Author: Fixed.

         If an earlier submission was accepted, you will eventually receive a :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>` rejection.

> 	 Otherwise, you have a second change to get the ledger change accepted on the ledger and learn the outcome.

Reviewer: ```suggestion
	 Otherwise, you have a second chance to get the ledger change accepted on the ledger and learn the outcome.
```

Author: Fixed.

	 

       - When you use the :ref:`Command Completion Service <command-completion-service>`, look for a completion for ``existing_submission_id`` instead of the chosen submission ID in :ref:`step 4 <dedup-bounded-step-await>`.

   - * ``ABORTED`` / other error codes
     

>      * Wait a bit and retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`).

Reviewer 3: "retry from Step 2" as above.

       
   - * other error conditions

>      * You should use background knowledge about the business workflow and the current ledger state to decide whether earlier submissions might still get accepted.

Reviewer 3: Delete "You should".

Author: Done.

>        If not, you may stop retrying and report that the ledger change failed.

Reviewer 3: Delete "you may".

Author: Done.

       In the running example of creating a contract instance of ``T``, you can never be sure, as any outstanding submission might still be accepted on the ledger.
       In particular, you must not draw any conclusions from not having received a :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>` error, because the outstanding submission may be queued somewhere and will reach the relevant processing point only later.

Failure scenarios
-----------------

The above strategy can fail in the following scenarios:

#. The bound ``B`` is too low: The command can be executed multiple times.
   
   Possible causes:

   - You have retried for longer than the deduplication duration, but never got a meaningful answer, e.g., because the timeout (gRPC deadline) is too short.
     For example, this can happen due to long-running Daml interpretation when using the :ref:`Command Service <command-service>`.

   - The application clock drifts significantly from the participant's or ledger's clock.

   - There are unexpected network delays.

   - Submissions are retried internally in the participant or Daml ledger and those retries do not stop before ``B`` is over.

>      Refer to the specific ledger's documentation for more information on such behaviour.

Reviewer 3: "on such behavior" can be deleted.

Author: Removed.

.. Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _command-deduplication:

Command deduplication
#####################

The interaction of a Daml application with the ledger is inherently asynchronous: applications send commands to the ledger, and some time later they see the effect of that command on the ledger.
Many things can fail during this time window:

- The application can crash.
- The participant node can crash.

> - Messages can be lost on the network

Author: Fixed.

.. note::
   The maximum deduplication time is the length of the deduplication period guaranteed to be supported by the participant.
   
The deduplication period chosen by the ledger is the *effective deduplication period*.
The ledger may also convert a requested deduplication duration into an effective deduplication offset or vice versa.
The effective deduplication period is reported in the command completion event in the :ref:`deduplication duration <com.daml.ledger.api.v1.Completion.deduplication_duration>` or :ref:`deduplication offset <com.daml.ledger.api.v1.Completion.deduplication_offset>` fields.

A command submission is considered a **duplicate submission** if at least one of the following holds:

- The submitting participant's completion service contains a successful completion event for the same :ref:`change ID <change-id>` within the *effective* deduplication period.

- The participant or Daml ledger are aware of another command submission in-flight with the same :ref:`change ID <change-id>` when they perform command deduplication.

The outcome of command deduplication is communicated as follows:

> - Command submissions via the :ref:`command service <command-service>` indicate the command deduplication outcome as a synchronous gRPC response unless when the `gRPC deadline <https://grpc.io/blog/deadlines/>`_ was exceeded.

Author: Fixed.

  .. note::
     The outcome MAY additionally appear as a completion event on the :ref:`command completion service <command-completion-service>`,
     but applications using the :ref:`command service <command-service>` typically need not process completion events.

- Command submissions via the :ref:`command submission service <command-submission-service>` can indicate the outcome as a synchronous gRPC response,
  or asynchronously through the :ref:`command completion service <command-completion-service>`.
  In particular, the submission may be a duplicate even if the command submission service acknowledges the submission with the gRPC status code ``OK``.

Independently of how the outcome is communicated, command deduplication generates the following outcomes of a command submission:

> - If there is no conflicting submission with the same :ref:`change ID <change-id>` on the Daml ledger or in-flight, the completion event and possibly the response convey the result of the submission (success or a gRPC error).

Reviewer 4: Shall we link the page for [gRPC status codes](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) here? I.e., it may be helpful to have the context around what the gRPC status codes below mean.

Author: I'll link to our own page about error codes, which also explains the gRPC status codes and the Daml error codes, along with a link to the official page.

      If in doubt, resubmit the commands using the same command ID.
      If the commands had been submitted before the crash, command deduplication on the ledger will reject the resubmissions.

   .. _dedup-bounded-step-offset:

#. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.

   .. _dedup-bounded-step-submit:
   
#. Submit the command with the following parameters:

   - Set the :ref:`command ID <<com.daml.ledger.api.v1.Commands.command_id>>` to the chosen command ID from :ref:`Step 1 <dedup-bounded-step-command-id>`.

   - Set the :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` to the bound ``B``.

   - Set the :ref:`submission ID <com.daml.ledger.api.v1.Commands.submission_id>` to a fresh value, e.g., a random UUID.

   - Set the timeout (gRPC deadline) to the expected submission processing time (Command Service) or submission hand-off time (Command Submission Service).

>      The **submission processing time** is time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives (synchronously, unless it times out) the acceptance or rejection.

Reviewer 4: ```suggestion
     The **submission processing time** is the time between when the application sends off a submission to the :ref:`Command Service <command-service>` and when it receives (synchronously, unless it times out) the acceptance or rejection.
```

Author: Fixed.

   
   - Status codes other than ``OK`` should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.

   - When you use the :ref:`Command Service <command-service>` and the response carries the status code ``OK``, the ledger change took place.
     You can report success.
     
   - When you use the :ref:`Command Submission Service <command-submission-service>`,
     subscribe with the :ref:`Command Completion Service <command-submission-service>` for completions for ``actAs`` from ``OFF1`` (exclusive) until you see a completion event for the change ID and the submission ID chosen in :ref:`Step 3 <dedup-bounded-step-submit>`.
     If the completion’s status is ``OK``, the ledger change took place and you can report success.
     Other status codes should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.
   
     This step needs no timeout as the :ref:`Command Submission Service <command-submission-service>` acknowledges a submission only if there will eventually be a completion event, unless relevant parts of the system become permanently unavailable.

.. _dedup-bounded-error-handling:

Error handling
--------------

> Error handling is needed when the status code of the command submission RPC call or in the :ref:`in the completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.

Reviewer 4: ```suggestion
Error handling is needed when the status code of the command submission RPC call or in the :ref:`completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.
```

Author: Fixed.

The following table lists appropriate reactions by status code (written as ``STATUS_CODE``) and error code (written in capital letters with a link to the error code documentation).
Fields in the error metadata are written as ``field`` in lowercase letters.

.. list-table:: Command deduplication error handling with known processing time bound
   :widths: 10 50
   :header-rows: 1

   - * Error condition
     
     * Reaction

       
   - * ``DEADLINE_EXCEEDED``
     
     * Consider the submission lost.
       

>        Retry from :ref:`Step 2 <dedup-bounded-step-offset>`, "obtaining the completion offset ``OFF1``", and possibly increase the timeout.

Reviewer 4: The exact text "obtaining the completion offset" is not part of `Step 2`, so this may be confusing. Maybe summarize the step in parenthesis?
```suggestion
       Retry from :ref:`Step 2 <dedup-bounded-step-offset>` (i.e., obtaining the completion offset ``OFF1``), and possibly increase the timeout.
```

Author: I had parentheses before and Carrie suggested to replace them by commas. As a compromise, I'll remove the quotes.

       
   - * Application crashed
     
     * Retry from :ref:`Step 2 <dedup-bounded-step-offset>`, "obtaining the completion offset ``OFF1``".

   - * ``ALREADY_EXISTS`` / :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>`
     
     * The change ID has already been accepted by the ledger within the reported deduplication period.
       The optional field ``completion_offset`` contains the precise offset.
       The optional field ``existing_submission_id`` contains the submission ID of the successful submission.
       Report success for the ledger change.
       

>        If desired, query the ``completion_offset`` via the :ref:`Command Completion Service <command-submission-service>` to find out about the earlier outcome.

Reviewer 5: It's currently not possible to query a specific `completion_offset`, and if we use the streaming approach the offset is actually exclusive. How would querying work in this case?

Author: Good point. I've removed this now. We can add it back in when/if we support querying by offset.

Offsets are arguably less intuitive and require more effort by the application developer.
We recommend the following strategy for using deduplication offsets:

#. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
   Remember the command ID across application crashes.
   (Analogous to :ref:`Step 1 above <dedup-bounded-step-command-id>`)

   .. _dedup-unbounded-step-dedup-offset:
   
#. Obtain a recent offset ``OFF0`` on the completion event stream and remember across crashes that you use ``OFF0`` with the chosen command ID. There are several ways to do so:

   - Use the :ref:`Command Service <command-service>` to obtain a recent offset by repeatedly submitting a dummy command, e.g., a :ref:`Create-And-Exercise command <com.daml.ledger.api.v1.CreateAndExerciseCommand>` of some single-signatory template with the :ref:`Archive <function-da-internal-template-functions-archive-52202>` choice, until you get a successful response.
     The response contains the :ref:`completion offset <com.daml.ledger.api.v1.SubmitAndWaitForTransactionIdResponse.completion_offset>`.

>    - Use the :ref:`Command Completion Service <command-completion-service>` by asking for the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.

Reviewer 4: From a practical point of view, the latter point is simpler to implement so I'd suggest reversing the order of these options for obtaining a recent offset.

Author: Reversed.

     If the completion’s status is ``OK``, the ledger change took place and you can report success.
     Other status codes should be handled according to :ref:`error handling <dedup-bounded-error-handling>`.

.. _dedup-unbounded-error-handling:

Error handling
--------------

The same as :ref:`for known bounds <dedup-bounded-error-handling>`, except that the former retry from :ref:`Step 2 <dedup-bounded-step-offset>` becomes retry from :ref:`Step 3 <dedup-unbounded-step-offset>`.

Failure scenarios
-----------------

The above strategy can fail in the following scenarios:

#. No success within the supported deduplication period
   
   When the application receives a :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>` error, it cannot achieve exactly once execution any more within the originally intended deduplication period.

#. Unacceptable changes cause infinite retries

   You need business workflow knowledge to decide that retrying does not make sense any more.
   Of course, you can always stop retrying and accept that you do not know the outcome for sure.

> .. todo:: 

Reviewer 4: Do you want to add this as part of a separate PR?

Author: AFAIK this is not yet implemented in the JSON API server. So there's no point yet in trying to document it. Whether that's going to be me or the person that implements it in the JSON API has not yet been decided.
